src/main/java/org/broadinstitute/hellbender/tools/dragstr/CalibrateDragstrModel.java

```diff
@@ -473,27 +480,120 @@ private StratifiedDragstrLocusCases collectCaseStatsParallel(final List<SimpleIn
-                                        throw new GATKException("problems accessing the str-table-file contents at " + strTablePath, ex);
+            return Utils.runInParallel(Math.min(threads, shards.size()), () ->
+                    StreamSupport.stream(new InterleavingListSpliterator<>(shards), true)
+                            .map(shard -> {
+                                try (final BinaryTableReader<DragstrLocus> lociReader = strTable.locusReader(shard)) {
+                                    final ReadsPathDataSource readsSource = threadReadSource.get();
+                                    final StratifiedDragstrLocusCases result = streamShardCasesStats(shard, readStream(readsSource, shard), lociReader.stream())
+                                            .collect(DragstrLocusCaseStratificator.make(hyperParameters.maxPeriod, hyperParameters.maxRepeatLength));
+                                    final int resultSize = result.size();
+                                    synchronized (numberBasesProcessed) {
+                                        final long processed = numberBasesProcessed.updateAndGet(l -> l + shard.size());
+                                        progressMeter.update(absoluteCoordinates.toSimpleInterval(processed, 1), resultSize);
                                     }
-                                }).reduce(StratifiedDragstrLocusCases::merge).orElseGet(() -> new StratifiedDragstrLocusCases(hyperParameters.maxPeriod, hyperParameters.maxRepeatLength)));
+                                    return result;
+                                } catch (final IOException ex) {
+                                    throw new GATKException("problems accessing the str-table-file contents at " + strTablePath, ex);
+                                }
+                            })
+                            .reduce(StratifiedDragstrLocusCases::merge)
+                            .orElseGet(() -> new StratifiedDragstrLocusCases(
+                                    hyperParameters.maxPeriod,
+                                    hyperParameters.maxRepeatLength)));
+        }
+    }
+
+    // Spliterator that splits by segragating even and odd elements as supposed to
+    // first and second half which the standard (e.g {@link Collection#stream} delgates onto {@link ArraySpliterator}).
+    //
+    // This approach makes the load more balanced between threads and the excecution order follows more closelly
+    // the input source list order.
+    //
+    // For its single use in this tool, this makes it more likely that there won't be a need to hold
+    // the full genome in memory by the underlaying ReferenceSource and that the
+    // actual progress corresponds more closely to the one reported by the progress-meter (always following
+    // sreference coordinate order).
+    private static class InterleavingListSpliterator<T> implements Spliterator<T> {
```

Comment:
It is tru that this could happen at the border of shards (1_000_000 bp long by default) but only if the adjacent ends are processed simultaneously since within the tool we only keep strongly referenced reads that overlap the sliding window of STR sites which should be roughly 2x  read-length. so around 500bp for a 250x2 dataset. 

Just to make it clear we don't hold the entire collection of read in shard in memory, but if the readers (bam/cram iterators) do that would be a problem .... I don't think that is the case, right?

So is possible but likely to be rare and even and if it happens most likely is inconsequential.

Reply:
As for the problem of holding the reference in memory....  Both using the standard or interleaved spliterator there is no need to hold a copy per thread;  the reason why that was happening is because I  was creating a separate ReferenceSource per reader and that reader were shared between threads thru a pool so in the end each reader/reference source pair would end up visiting the genome at random ....

That is fixed in the new code in two ways, now the readers are attach to their respective threads and on top of that they share the sane reference-source. The second change is actually not that necessary if you were to keep using the regular spliterator. Using the interleaved one you save reading the reference genome N times as every thread goes thru it.  

The classical spliterator would tend to keep in memory N contigs where N is the number of threads, the interleaved it would be less than N; 1 at the begining as everybody starts at chr1 and as it goes along the genome it would potentially increase up to N as the threads performance may vary but since contigs get smaller and smaller that is less of a problem the further it goes ... so one may argue is the O(size of the largest contig) for human ref at least.